### DISCUSSION TITLE
Data Importer: Duplicate when transferring between accounts

### DISCUSSION THREAD

**Discussion by NigelVanHattum** [firefly-iii/firefly-iii#8463] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

When I use the bulk export from ING to bulk import all my transactions. The transfers between my 2 ING accounts get doubled. My CSV is added here as an example (Removed ignored fields):

| Date         | Name/ Description     | Account | Counterparty | Debit/Credit | Amount |
|--------------|-----------|------------|------------|------------|------------|
| 20240125 |  Example     | My first asset account  | My second asset account      | credit | **  |

ING only creates one row for a transfer from 1 account to the other. 

However, when I look at Firefly it created the Transfer 2 times. 
| Date         | Name/ Description     | Account | Counterparty | Debit/Credit | Amount |
|--------------|-----------|------------|------------|------------|------------|
| 20240125 |  Example     | My first asset account | My second asset account      | credit | **  |
| 20240125 |  Example     | My first asset account | My second asset account      | credit | **  |

Looking at the documentation it clearly states that "Each journal contains two "transactions", but in this case this is not what I would expect. This now causes both my asset accounts to double their transaction values, meaning My first asset account receives a double negative transaction value and my second a double positive. 

### Debug information

Debug information generated at 2024-01-26 13:05:14 for Firefly III version **v6.1.7**.

<table><thead><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Firefly III</td><td>v6.1.7 / v2.0.12 / 22 (exp. 22)</td></tr><tr><td>PHP version</td><td>8.3.2 / apache2handler / Linux</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Error reporting</td><td>Display: Off, reporting: ALL errors</td></tr><tr><td>Max upload</td><td>67108864 (64 MB)</td></tr><tr><td>Database drivers</td><td>mysql, <strong>*pgsql*</strong>, sqlite, </td></tr><tr><td>Docker build</td><td><span>#</span>933, base <span>#</span>79</td></tr></tbody></table><table><thead><tr><th colspan="2">Firefly III information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Timezone</td><td>Europe/Amsterdam + Europe/Amsterdam</td></tr><tr><td>App environment</td><td>production, debug: false</td></tr><tr><td>Layout</td><td>v1</td></tr><tr><td>Logging</td><td>info, stack / (empty)</td></tr><tr><td>Cache driver</td><td>file</td></tr><tr><td>Default language and locale</td><td>en_US + equal</td></tr><tr><td>Trusted proxies</td><td>**</td></tr><tr><td>Login provider & user guard</td><td>eloquent / web</td></tr><tr><td>Login headers</td><td>N/A + N/A</td></tr><tr><td>Stateful domains</td><td></td></tr><tr><td>Last cron job</td><td>never (never)</td></tr><tr><td>Mailer</td><td>log</td></tr></tbody></table><table><thead><tr><th colspan="2">User-specific information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>User</td><td><span>#</span>1 of 1</td></tr><tr><td>User flags</td><td><span title="Has rules">:wrench:</span></td></tr><tr><td>Session start</td><td>2023-12-01 12:26:25</td></tr><tr><td>Session end</td><td>2023-12-31 12:26:25</td></tr><tr><td>View range</td><td>1M</td></tr><tr><td>User language</td><td>en_US</td></tr><tr><td>User locale</td><td>en_US</td></tr><tr><td>Locale(s) supported</td><td>en_US.utf8: :white_check_mark:<br>en_US.UTF-8: :white_check_mark:<br></td></tr><tr><td>User agent</td><td>Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 Edg/120.0.0.0</td></tr></tbody></table>

### Expected behaviour

I would expect that a transfer between 2 of my asset accounts only registers once. 

### Steps to reproduce

1. Export ING transactions via the UI, exporting all accounts to a single CSV file. (or 2 separate CSV's, this does not matter)
2. Import the CSV via the firefly-importer
3. Using the following config:
    - Headers: true
    - CSV-file delimiter: A semicolon (;)
    - Default import account: Cash Wallet 
    - Detection method: content based
4. Role config
    - Date: Transaction Date
    - Name/ Description: Description
    - Account: Asset Account (IBAN)
    - Counterparty: Opposing account (IBAN)
    - Debit/credit: ING specific debit/credit indicator
    - Amount (EUR): amount
    - Notifications: Note(s)

**Comment by JC5** [comment]
This happens when the two rows aren't identical (description, date, amount, account numbers, etc.). It should be easy to see in the CSV file.

**Comment by NigelVanHattum** [comment]
As I mentioned, when I use a single CSV file. It only contains one row for the transfer transaction. 

Look at the 2 different CSV's Only the account from which the transaction origins changes. Because the export is for the other account. 

| Date         | Name/ Description     | Account | Counterparty | Debit/Credit | Amount | Note(s)|
|--------------|-----------|------------|------------|------------|------------|------------|
| 20240125 |  Example     | My first asset account | My second asset account      | credit | **  | Example note
| 20240125 |  Example     | My second asset account | My first asset account      | debit | **  | Example note

**Comment by JC5** [comment]
https://docs.firefly-iii.org/how-to/data-importer/import/duplicates/

See "debugging".

**Comment by NigelVanHattum** [comment]
I found out that the "note(s)" create a different hash, but without the note(s) column it filters out payments made through "Tikkie" since they are also the exact same amount, but only differ in note

**Comment by dlegranddk** [comment]
I can maybe give fresh inputs here, I had to battle a bit between the "auto-transfers" (recurring transfers I programmed at the bank level) and the "manual" ones (transfers i would do on the spot on my bank mobile app).

Basically, even if the import tool is great, the data provided by some banks isn't always that good and the import tool shouldn't try to guess too much, so I ended up trusting only the description and to use rules instead with a clear logic:
1) Delete the credit transaction where the description contains "from myself xxxx"
2) Convert the debit transaction where the description contains "to: theTargetBankAccount"

So if you see data issue with your bank, check what your bank provides properly and create a dedicated rule group, easier to keep an overview, specially when you have a lot of accounts, and be consistent like delete the credit line and update the debit line.
1) If the data provided by the bank allows the import tool to find both source/target accounts, just delete one line and convert if necessary the other one
2) If the data isn't good so you have a credit with no source and a debit with no target, then use the description like I try to do


![Screenshot 2024-01-31 140017](https://github.com/firefly-iii/firefly-iii/assets/104696958/71b765ae-dfc1-4523-8179-55a2464bd30a)

And as the rule engine can't use Regular Expressions, I would recommend to adopt also a convention for the "convert to transfer rule", I now use a convention where one condition is a fixed description all transactions will have and a second one which is the target bank account number.
![Screenshot 2024-01-31 144236](https://github.com/firefly-iii/firefly-iii/assets/104696958/7b012d86-b1a4-4ffe-94d0-126c10acf979)

It's my 3rd end of the month, and 3rd attempt to get this automated, I still got some duplicates but hopefully, I'm closer to the right formula for me :) and I can foresee that the time spent at figurring out the rules, and the transfer ones are the most complex, the other ones being easier to manage will be paid back in time saved at budgetting.

I hope it helps.
